### ISSUE TITLE
raft topology: assigning tokens races with `setup_group0`, causing data path errors

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#15935]
In the new node joining procedure (https://github.com/scylladb/scylladb/pull/15196), the topology coordinator assigns tokens to the joining node when moving out of `join_group0` transition state. Concurrently, the joining node is inside `setup_group0` -- after joining group 0, it enters `group0_upgrade_state::synchronize`, performs a `read_barrier`, and then moves into `group0_upgrade_state::use_post_raft_procedures` after confirming that the cluster is ready to use Raft.

This is problematic, because while in `group0_upgrade_state::synchronize`, the node might get tokens assigned already and start receiving writes. These writes might use schema which the node doesn't have yet (it may still be inside the `read_barrier`, applying latest group 0 commands), so they trigger `get_schema_for_write`, which triggers `start_operation()`, which detects that we're inside `synchronize` and throws an error. So the writes fail during this time.

We should delay the step of assigning tokens until after the node finished synchronizing with the latest group 0 commands and entered `use_post_raft_procedures`.

cc @piodul @gleb-cloudius

**Comment by kostja** [comment]
If the new topology coordinator is used, upgrade is pointless  and should be skipped.

**Comment by piodul** [comment]
The `join_node_response_handler` already performs a read barrier, causing the current schema to be loaded. Only after the handler finishes, the topology coordinator assigns tokens to the new node - same goes for entering `synchronize` and later `use_post_raft_procedures`. Therefore, the node should have the most recent schema when it handles requests while in `synchronize`.

AFAIK the issue was opened after observing failures in https://github.com/scylladb/scylladb/pull/15890#issuecomment-1801975517. The changes introduced there caused schema mismatch to happen between nodes (for more details see here: https://github.com/scylladb/scylladb/issues/16004), and most likely is the reason why the joining node tried to pull schema when handling writes in `synchronize` state.

Naturally, there is a possibility that a schema change happens while the joining node is in `synchronize` state and it will try to pull. However, it doesn't happen in the `test_topology_ops`, so it shouldn't be a blocker for reenabling the test (https://github.com/scylladb/scylladb/issues/15962). Fixing this right now is possible and not difficult (by doing what Kostja suggests), but the fix would most likely have to be adjusted while developing the upgrade procedure (https://github.com/scylladb/scylladb/issues/15008), so it probably makes sense to postpone the fix and do it along with the upgrade procedure.

**Comment by piodul** [comment]
Note: I've added a 3 second sleep after the node enters `synchronize` in order to prolong it, ran the test 100 times locally in dev mode on `master` (with writes re-enabled) and confirm that the errors related to failed schema pulls due to the `synchronize` state doesn't occur.

**Comment by kbr-scylla** [comment]
Ok, thanks. I don't think we need this issue. This upgrade path will probably get significant modifications when implementing upgrade procedure for raft-topology (it's already in a sense "broken" for raft-topology mode) so the issue does not "apply".
